File: libs/flutter/android/src/main/java/com/reblaze/ReblazeFlutterPlugin.java
package com.reblaze;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

import static com.reblaze.sdk.Reblaze.reblaze;

import android.util.Log;

import java.util.List;

/** ReblazePlugin */
public class ReblazeFlutterPlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "reblaze.com");
    channel.setMethodCallHandler(this);
    if (BuildConfig.DEBUG)
    reblaze.setEventListener((kind, s) -> {
      Log.println(kind.getValue(), "flutter", s);
    });
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    switch (call.method) {
      case "generateHash":
        result.success(reblaze.generateHash());
        return;
      case "sendEvent":
        String eventName = (String) call.arguments();
        if (eventName != null) {
          reblaze.sendEvent(eventName);
        }
        return;
      case "setBackendUrl":
        String backendUrl = (String) call.arguments();
        if (backendUrl != null) {
          reblaze.setBackendUrl(backendUrl);
        }
        return;
      case "getBackendUrl":
        result.success(reblaze.getBackendUrl());
        return;
      case "getUserAgent":
        result.success(reblaze.getUserAgent());
        return;
      case "getConfigurationName":
        result.success(reblaze.getConfigurationName());
        return;
      case "getHashVariant":
        result.success(reblaze.getHashVariant());
        return;
      case "getToken":
        result.success(reblaze.getToken());
        return;
      case "getTokenHeaderName":
        result.success(reblaze.getTokenHeaderName());
        return;
      case "getTimeslice":
        result.success(reblaze.getTimeslice());
        return;
      case "getInterval":
        result.success(reblaze.getInterval());
        return;
      case "getReportCounters":
        result.success(reblaze.getReportCounters());
        return;
      case "setToken":
        String token = (String) call.arguments();
        if (token != null) {
          reblaze.setToken(token);
        }
        return;
      case "setTokenHeaderName":
        String tokenHeaderName = (String) call.arguments();
        if (tokenHeaderName != null) {
          reblaze.setTokenHeaderName(tokenHeaderName);
        }
        return;
      case "setTimeslice":
        Integer timeslice = (Integer) call.arguments();
        if (timeslice != null) {
          reblaze.setTimeslice(timeslice);
        }
        return;
      case "setInterval":
        Integer interval = (Integer) call.arguments();
        if (interval != null) {
          reblaze.setInterval(interval);
        }
        return;
      case "setReportCounters":
        String reportCounters = (String) call.arguments();
        if (reportCounters != null) {
          reblaze.setReportCounters(reportCounters.equals(com.reblaze.sdk.ReportCounters.ALL.name()) ? com.reblaze.sdk.ReportCounters.ALL : com.reblaze.sdk.ReportCounters.NONE);
        }
        return;
      case "enableMock":
        reblaze.enableMock();
        return;
      case "setMockResponse":
        List<Object> args = (List<Object>) call.arguments();
        if (args != null && args.size() == 3) {
          Integer code = (Integer) args.get(0);
          String remoteConfig = (String) args.get(1);
          Integer delay = (Integer) args.get(2);
          if (delay == -1) {
            reblaze.setMockResponse(code, remoteConfig);
          }
          else {
            reblaze.setMockResponse(code, remoteConfig, delay);
          }
        }
        else {
          result.notImplemented();
        }
        return;
      default:
        result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    reblaze.setEventListener(null);
    channel.setMethodCallHandler(null);
  }

}

File: libs/flutter/ios/Classes/ReblazeFlutterPlugin.m
#import "ReblazeFlutterPlugin.h"

@import ReblazeSDK;

@implementation ReblazeFlutterPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"reblaze.com"
            binaryMessenger:[registrar messenger]];
  ReblazeFlutterPlugin* instance = [[ReblazeFlutterPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([call.method isEqualToString: @"generateHash"]) {
    result([reblaze generateHash]);
  } else if ([call.method isEqualToString: @"sendEvent"]) {
    [reblaze sendEvent:call.arguments];
  } else if ([call.method isEqualToString: @"setBackendUrl"]) {
    reblaze.backendUrl = call.arguments;
  } else if ([call.method isEqualToString: @"getBackendUrl"]) {
    result(reblaze.backendUrl);
  } else if ([call.method isEqualToString: @"getUserAgent"]) {
    result(reblaze.userAgent == nil ? @"" : reblaze.userAgent);
  } else if ([call.method isEqualToString: @"getConfigurationName"]) {
    result(reblaze.configurationName);
  } else if ([call.method isEqualToString: @"getHashVariant"]) {
    result(reblaze.hashVariant);
  } else if ([call.method isEqualToString: @"getToken"]) {
    result(reblaze.token);
  } else if ([call.method isEqualToString: @"getTokenHeaderName"]) {
    result(reblaze.tokenHeaderName);
  } else if ([call.method isEqualToString: @"getTimeslice"]) {
    result([NSString stringWithFormat:@"%ld", (long)reblaze.timeslice]);
  } else if ([call.method isEqualToString: @"getInterval"]) {
    result([NSString stringWithFormat:@"%ld", (long)reblaze.interval]);
  } else if ([call.method isEqualToString: @"getReportCounters"]) {
    result(reblaze.reportCounters == ReblazeSDK_ReportCountersNONE ? @"NONE" : @"ALL");
  } else if ([call.method isEqualToString: @"setToken"]) {
    reblaze.token = call.arguments;
  } else if ([call.method isEqualToString: @"setTokenHeaderName"]) {
    reblaze.tokenHeaderName = call.arguments;
  } else if ([call.method isEqualToString: @"setTimeslice"]) {
    reblaze.timeslice = [call.arguments intValue];
  } else if ([call.method isEqualToString: @"setInterval"]) {
    reblaze.interval = [call.arguments intValue];
  } else if ([call.method isEqualToString: @"setReportCounters"]) {
    reblaze.reportCounters = ([call.arguments isEqualToString: @"NONE"] ? ReblazeSDK_ReportCountersNONE : ReblazeSDK_ReportCountersALL);
  } else if ([call.method isEqualToString: @"enableMock"]) {
    [reblaze enableMock];
  } else if ([call.method isEqualToString: @"setMockResponse"]) {
      NSArray *args = (NSArray *)call.arguments;
      if (args != nil && args.count == 3) {
          int code = (int) args[0];
          NSString *remoteConfig = (NSString *) args[1];
          int delay = (int) args[2];
          if (delay == -1) {
              [reblaze setMockResponse:code remoteConfig:remoteConfig];
          }
          else {
              [reblaze setMockResponse:code remoteConfig:remoteConfig delay:delay];
          }
      }
  } else if ([call.method isEqualToString: @"setAutoSign"]) {
    reblaze.reportCounters = ([call.arguments isEqualToString: @"NONE"] ? ReblazeSDK_AutoSignNONE : ReblazeSDK_AutoSignALL);
  } else if ([call.method isEqualToString: @"getAutoSign"]) {
    result(reblaze.reportCounters == ReblazeSDK_AutoSignNONE ? @"NONE" : @"ALL");
  } else {
    result(FlutterMethodNotImplemented);
  }
}

@end
